lib/spack/spack/audit.py

```diff
@@ -282,18 +282,35 @@ def _search_duplicate_specs_in_externals(error_cls):
 
 
 @package_directives
-def _check_build_test_callbacks(pkgs, error_cls):
-    """Ensure stand-alone test method is not included in build-time callbacks"""
+def _check_test_callbacks(pkgs, error_cls):
+    """Ensure stand-alone test method is not included in test callbacks."""
```

Comment:
Not sure changing the semantics of the test is what we want. I'll have a look at the issue originating the PR soon.

Reply:
1. The audit check as it existed previously was faulty as it _assumed_ that the callbacks were defined in every package class and broke the audit with an exception at the first package for which that was not true. See the originating issue for why solving this problem by defaulting the callbacks in `PackageBase` causes worse problems than it solves.
2. There seemed to be no obvious reason (other than that the audit check predated the advent of `install_time_test_callbacks`) for the presence of the user `test()` method in `build_time_test_callbacks` is bad, but in `install_time_test_callbacks` is not, which is why I extended the audit to include both callbacks.
3. If it is only necessary to exclude the package's `test()` method from legacy  callback lists, but having a builder `test()` method in builder callback lists is OK, I would be happy to adjust this PR accordingly.
4. The auditing question is entirely secondary to the issue described in #33928, for which I believe the correct fix remains the excisions implemented in [ec2f858](https://github.com/spack/spack/pull/33942/commits/ec2f8589c872b81925612b581cd4f9bf1e25aaae)

Reply:
I agree with 4. I cherry picked your commit in #33992 where I also added a regression test.

For 1-3 my understanding is that _there are_ reasons for that. Install time tests are meant to be executed in the prefix, and as such it might make sense to reuse the standalone test method. Build time tests instead are executed in the stage directory, so the standalone test method should not be called there. @tldahlgren can correct me if my understanding is wrong (I am not using the standalone tests a lot, so may be missing details).

Reply:
@alalazo Thanks for this. I understand your explanation of why install-time tests may use the standalone `test()` method, but I'm missing why there's no need to audit the user builder's `build_time_test_callbacks` in addition to the legacy attribute?

Reply:
https://github.com/spack/spack/pull/33942/commits/b7f788951c1770700a8c8dbc069f7f4b47787ac3 removes the audit of `install_time_test_callbacks` while preserving the audit of builder attributes in addition to the legacy attribute. I'd also be happy to bring in https://github.com/spack/spack/pull/33992/commits/bd80adb15b9019c1bce74658e3e2eea9dd16ba27 if you agree— I initially didn't understand exactly how one would test the regression efficiently, and was intending to discuss same as part of the PR.

Going forward, please know that I am always happy to revise PRs based on reviewer or other comments—even those suggesting minor or cosmetic changes—and take on board knowledge of why things are/aren't/should be/shouldn't be done. As a core developer, I'm sure you have more pressing demands on your time and expertise than re-implementing fixes to minor bugs.

Reply:
> As a core developer, I'm sure you have more pressing demands on your time and expertise than re-implementing fixes to minor bugs.

Apologies @chissg For this bug I cherry-picked the commit and submitted a new PR since the bug has a high impact on the builtin repository, so I could have a fast turnaround. In that sense it's not minor. 

Also this PR was missing a regression test and is also dealing with audits, which are a separate issue. If you want to remove any change to audit and cherry-pick https://github.com/spack/spack/commit/bd80adb15b9019c1bce74658e3e2eea9dd16ba27 here I am absolutely fine with that - the objective is to fix Spack, not to get _my_ PR in.

Reply:
> why there's no need to audit the user builder's build_time_test_callbacks in addition to the legacy attribute?

As far as I understand, the standalone tests are based on the `PackageBase` hierarchy and they do not need builders. Therefore any `test` method in the `Builder` will not interfere with `spack test`.

Reply:
Fair enough, thank you for the explanation.